File: python/tests/test_conformance_load.py
"""Golden load cases (shared tests/cases_load.csv) — must pass unmodified.

Row counts pin the merge semantics (person 1:1 outer chain, then hh m:1
outer keeping unmatched hh rows); the expect_windex8 column pins the
no-mutation contract (windex5 == 8 must SURVIVE loading — the pre-contract
Stata loader silently recoded it to missing).
"""

from __future__ import annotations

import pandas as pd
import pytest

from conftest import read_cases

import datalib as dl
from datalib.errors import InvalidArgumentError, NotFoundError

CASES = read_cases("cases_load.csv")


def _counts(result) -> tuple[int, int]:
    """(row count, windex5==8 count) over a DataFrame or a module dict."""
    frames = list(result.values()) if isinstance(result, dict) else [result]
    n = sum(len(df) for df in frames)
    w8 = sum(
        int((df["windex5"] == 8).sum())
        for df in frames
        if "windex5" in df.columns
    )
    return n, w8


@pytest.mark.parametrize("case", CASES, ids=[c["case_id"] for c in CASES])
def test_load_case(case: dict[str, str], fixture_root) -> None:
    kwargs = dict(
        country=case["country"],
        year=case["year"] or None,
        survey=case["survey"] or None,
        kind=case["kind"] or "adaptation",
        collection=case["collection"] or None,
        modules=case["modules"].split() or None,
        master_version=case["master_version"] or None,
        merge=not case["nomerge"],
        root=fixture_root,
    )
    if case["expect_error"]:
        with pytest.raises((InvalidArgumentError, NotFoundError)):
            dl.load(**kwargs)
        return

    result = dl.load(**kwargs)
    if case["nomerge"]:
        assert isinstance(result, dict)
    else:
        assert isinstance(result, pd.DataFrame)

    n, w8 = _counts(result)
    assert n == int(case["expect_n"]), f"row count for {case['case_id']}"
    if case["expect_windex8"]:
        assert w8 == int(case["expect_windex8"]), (
            f"windex5==8 count for {case['case_id']} (no-mutation contract)"
        )

    # provenance columns are always present after load
    frames = list(result.values()) if isinstance(result, dict) else [result]
    for df in frames:
        assert "ctrycode" in df.columns
        assert "year" in df.columns
        assert (df["ctrycode"] == case["country"].upper()).all()

File: r/tests/testthat/test-conformance-load.R
# Conformance: the SHARED golden load cases (tests/cases_load.csv at repo
# level) must pass unmodified. expect_n pins the merged (or per-module) row
# count; expect_windex8 pins the count of windex5 == 8 surviving the load
# (the no-mutation contract). nomerge means merge = FALSE (named list).

test_that("shared load conformance cases pass unmodified", {
  cases <- read.csv(file.path(datalib_test_repo_tests(), "cases_load.csv"),
                    colClasses = "character")
  expect_gt(nrow(cases), 0)
  root <- datalib_test_fixture_root()
  expect_true(dir.exists(root))

  nz <- function(x) !is.na(x) && nzchar(x)
  count_windex8 <- function(df) {
    if (!"windex5" %in% names(df)) return(0L)
    w <- as.numeric(df$windex5)
    as.integer(sum(!is.na(w) & w == 8))
  }

  for (i in seq_len(nrow(cases))) {
    cs <- cases[i, ]
    args <- list(country = cs$country, kind = cs$kind, root = root,
                 merge = !nz(cs$nomerge))
    if (nz(cs$year)) args$year <- cs$year
    if (nz(cs$survey)) args$survey <- cs$survey
    if (nz(cs$collection)) args$collection <- cs$collection
    if (nz(cs$master_version)) args$master_version <- cs$master_version
    if (nz(cs$modules)) {
      args$module <- strsplit(trimws(cs$modules), "\\s+")[[1]]
    }

    if (nz(cs$expect_error)) {
      expect_error(do.call(datalib_load, args),
                   class = paste0("datalib_error_", cs$expect_error),
                   info = paste("case", cs$case_id))
      next
    }

    res <- do.call(datalib_load, args)
    if (is.data.frame(res)) {
      n <- nrow(res)
      w8 <- count_windex8(res)
    } else {
      expect_true(is.list(res), info = paste("case", cs$case_id))
      n <- sum(vapply(res, nrow, integer(1)))
      w8 <- sum(vapply(res, count_windex8, integer(1)))
    }
    expect_identical(as.integer(n), as.integer(cs$expect_n),
                     info = paste("case", cs$case_id, "row count"))
    if (nz(cs$expect_windex8)) {
      expect_identical(as.integer(w8), as.integer(cs$expect_windex8),
                       info = paste("case", cs$case_id, "windex5==8 count"))
    }
    if (nz(cs$expect_provenance) && cs$expect_provenance == "1") {
      frames <- if (is.data.frame(res)) list(res) else res
      for (f in frames) {
        expect_true(all(c("ctrycode", "year") %in% names(f)),
                    info = paste("case", cs$case_id, "provenance columns"))
      }
    }
  }
})
